src/ert/run_models/model_factory.py

```diff
@@ -158,18 +164,46 @@ def _setup_evaluate_ensemble(
     )
 
 
-def _get_active_realizations_list(args: Namespace, config: ErtConfig) -> list[bool]:
+def _get_and_validate_active_realizations_list(
+    args: Namespace, config: ErtConfig
+) -> list[bool]:
     ensemble_size = config.runpath_config.num_realizations
     if (
         config.analysis_config.design_matrix is not None
-        and config.analysis_config.design_matrix.active_realizations is not None
+        and (
+            dm_active_realizations
+            := config.analysis_config.design_matrix.active_realizations
+        )
+        is not None
     ):
-        if args.realizations is None:
-            return config.analysis_config.design_matrix.active_realizations
-        else:
-            ensemble_size = len(
-                config.analysis_config.design_matrix.active_realizations
+        if ensemble_size != len(dm_active_realizations):
+            ensemble_size = min(ensemble_size, len(dm_active_realizations))
+            logger.warning(
+                "The number of realizations in the design matrix "
+                f"({len(dm_active_realizations)}) differs from the configured "
+                f"ensemble size ({config.runpath_config.num_realizations}). "
+                f"Using the minimum of both ({ensemble_size})."
             )
+        if hasattr(args, "realizations") and args.realizations is not None:
```

Comment:
you can shorten this to
`if getattr(args, "realizations", None) is not None:`
or perhaps
`if reals := getattr(args, "realizations", None):`
although this would also be false if realizations is a falsy value